src/content/blog/2020-12-07-roadmap-2021.md

```diff
@@ -0,0 +1,238 @@
+
+## What happened so far?
+
+As few more things has been exposed from webpack, both typings-wise and runtime-wise. A few low-handling performance improvements have been done. Code without semicolons used to generate some invalid/incorrect code in some cases, which has been fixed. The combination of side-effect-free code + concatenated modules + reexports has led to some edge cases, which have been fixed (at least the known ones).
+
+But one bug reported by an user has led to the need to a completely new internal feature.
+You can skip to the next chapter if you find webpack internals boring or too complex.
+
+To trigger the bug we need three ingredients:
+
+- Since webpack 5 the optimization in production mode will run the used exports analysis (Tree Shaking) for each runtime (which is often identical to the entrypoints). So webpack can optimize each runtime (resp. entrypoint) individually.
+- A custom `optimization.splitChunks` configuration allows to forcefully merge modules into a single chunk. This is done by passing the `name` option. E. g. `test: /node_modules/, name: "vendors"` merges all node_modules into a single chunk. While this is not recommended in general, it's possible and probably makes sense in some cases. The whole thing is about trade-offs anyway and choosing to merge all vendors into a single chunk can be good for long term caching this chunk for repeated visits or between mulitple entrypoints.
+- When exports of a side-effect-free module are not used, the whole module is omitted from the module graph and the `import`-statement generates no runtime code at all.
+
+A problem happens in the edge case when modules from two entrypoints are merged into a single chunk and these module reference a side-effect-free module which are not in the shared chunk.
+And only one the entrypoints use exports for the side-effect-free module.
+The modules in the shared chunk are used by both entrypoint, so they need to include the exports that are used by any of the entrypoints.
+In the mentioned edge case this will generate code which references the side-effect-free module which is not in the share chunk.
+But as the side-effect-free module is only needed by one of the entrypoints, it's only included in only of the entrypoint.
+This means the module is not available at runtime for the other entrypoint, which leads to an `undefined is not a function` or `cannot read property 'call' of undefined` error at runtime.
```

Comment:
```suggestion
A problem happens in an edge case where modules from two entrypoints are merged into a single chunk and they reference a side-effect-free module which is not in the shared chunk since only one of the entrypoints uses exports from the side-effect-free module. Modules in the shared chunk are used by both entrypoint, thus they need to include the exports that are used by any of the entrypoints. That means it will generate code referencing the side-effect-free module that is not available at runtime for the other entrypoint in the aforementioned edge case, which leads to an `undefined is not a function` or `cannot read property 'call' of undefined` error at runtime.
```
I feel we can improve these, but I'm not familiar with the internals so feel free to edit.